packages/styled/__tests__/as-prop.js

```diff
@@ -0,0 +1,125 @@
+// @flow
+import 'test-utils/next-env'
+import React from 'react'
+import { render, cleanup } from 'react-testing-library'
+import styled from '@emotion/styled'
+
+let id = () => Math.random().toString(36)
+
+afterEach(cleanup)
+
+test('string base - string as', () => {
+  const Title = styled('h1')`
+    color: green;
+  `
+
+  let h1Title = id()
+  let h2Title = id()
+
+  let { getByTestId } = render(
+    <article>
+      <Title data-testid={h1Title}>My Title</Title>
+      <Title data-testid={h2Title} as="h2">
+        My Subtitle
+      </Title>
+    </article>
+  )
+
+  expect(getByTestId(h1Title).tagName).toBe('H1')
+  expect(getByTestId(h2Title).tagName).toBe('H2')
+})
+
+test('styled string base - string as', () => {
+  const Title = styled('h1')`
+    color: green;
+  `
+
+  const Subtitle = styled(Title)`
+    font-weight: 100;
+  `
+
+  let h1Title = id()
+  let h1Subtitle = id()
+  let h3Subtitle = id()
+
+  let { getByTestId } = render(
+    <article>
+      <Title data-testid={h1Title}>My Title</Title>
+      <Subtitle data-testid={h1Subtitle}>My Subtitle</Subtitle>
+      <Subtitle data-testid={h3Subtitle} as="h3">
+        My smaller Subtitle
+      </Subtitle>
+    </article>
+  )
+
+  expect(getByTestId(h1Title).tagName).toBe('H1')
+  expect(getByTestId(h1Subtitle).tagName).toBe('H1')
+  expect(getByTestId(h3Subtitle).tagName).toBe('H3')
+})
+
+test('composite base - string as', () => {
+  const Title = styled(props => <h1 {...props} />)`
+    color: green;
+  `
+  let h1Title = id()
+  let h3Title = id()
+
+  let { getByTestId } = render(
+    <article>
+      <Title data-testid={h1Title}>My Title</Title>
+      <Title data-testid={h3Title} as="h3">
+        My Subtitle
+      </Title>
+    </article>
+  )
+
+  expect(getByTestId(h1Title).tagName).toBe('H1')
+  expect(getByTestId(h3Title).tagName).toBe('H3')
+})
```

Comment:
I disagree with this behavior. I think if you pass a composite base, we should forward `as` unless you explicitly pass a `shouldForwardProp`.

Reply:
I was not sure how to handle this case. The reasoning behind this solution is that most people don't use `shouldForwardProp` feature and this logic makes more common scenarios easier.

With string `as` here we kinda implicitly know which props should be forwarded regardless of the base type and it seems to me that this would be the expected outcome for most people. It also aligns the default behaviour to `styled-components` so this would make migrating from one lib to another easier.

I'm not overly attached to the logic here, so feel free to decide how this should work and I'll adjust the code.

Reply:
I feel like the default should handle the most common use case and I feel like the most common use case for passing a non string base is that it handles `as`, while there are certainly some use cases where styled should handle `as` for composite bases, I think it won't be as common.

Reply:
Yeah, choosing between compatibility with `styled-components` or (IMO) better behavior is hard. I feel like for now, let's make it so by default composite bases forwards as(unless a shouldForwardProp that doesn't forward as is passed) and see what people think and then go from there. I feel pretty confident with it since I feel like a more real world usecase would be more like this.
```jsx
const Link = styled.a``

<Link as={DifferentLink} />
```

Reply:
My take on this.

```
1. if emotion component forward `as`
2. if composite component forward `as`
3. if emotion is writing the final string tag use the `as` prop to overwrite the tag and *do not forward* `as`
4. if custom shouldForwardProp returns `false` for `'as'` *do not forward* `as`
```